src/Symfony/Component/Messenger/Worker.php

```diff
@@ -32,112 +32,141 @@
-
-            $message = $envelope->getMessage();
-            $context = [
-                'message' => $message,
-                'class' => \get_class($message),
-            ];
-
-            try {
-                $envelope = $this->bus->dispatch($envelope->with(new ReceivedStamp()));
-            } catch (\Throwable $throwable) {
-                $shouldRetry = $this->shouldRetry($throwable, $envelope);
-
-                $this->dispatchEvent(new WorkerMessageFailedEvent($envelope, $this->receiverName, $throwable, $shouldRetry));
-
-                if ($shouldRetry) {
-                    if (null === $this->retryStrategy) {
-                        // not logically allowed, but check just in case
-                        throw new LogicException('Retrying is not supported without a retry strategy.');
-                    }
-
-                    $retryCount = $this->getRetryCount($envelope) + 1;
-                    if (null !== $this->logger) {
-                        $this->logger->info('Retrying {class} - retry #{retryCount}.', $context + ['retryCount' => $retryCount, 'error' => $throwable]);
-                    }
-
-                    // add the delay and retry stamp info + remove ReceivedStamp
-                    $retryEnvelope = $envelope->with(new DelayStamp($this->retryStrategy->getWaitingTime($envelope)))
-                        ->with(new RedeliveryStamp($retryCount, $this->getSenderAlias($envelope)))
-                        ->withoutAll(ReceivedStamp::class);
-
-                    // re-send the message
-                    $this->bus->dispatch($retryEnvelope);
-                    // acknowledge the previous message has received
-                    $this->receiver->ack($envelope);
-                } else {
-                    if (null !== $this->logger) {
-                        $this->logger->info('Rejecting {class} (removing from transport).', $context + ['error' => $throwable]);
-                    }
-
-                    $this->receiver->reject($envelope);
-                }
+        while (false === $this->shouldStop) {
+            $envelopeHandled = false;
+            foreach ($this->receivers as $receiverName => $receiver) {
+                $envelopes = $receiver->get();
+
+                foreach ($envelopes as $envelope) {
+                    $envelopeHandled = true;
 
-                if (\function_exists('pcntl_signal_dispatch')) {
-                    pcntl_signal_dispatch();
+                    $this->handleMessage($envelope, $receiver, $receiverName, $this->retryStrategies[$receiverName] ?? null);
+                    $handled($envelope);
                 }
 
-                return;
+                // after handling a single receiver, quit and start the loop again
+                // this should prevent multiple lower priority receivers from
+                // blocking too long before the higher priority are checked
+                if ($envelopeHandled) {
```

Comment:
Important decision: once we get at least one message from one receiver, we immediately go back to the beginning of the loop. The idea is that if you have `high`, `medium`, `low` transports, if there are none on `high`, but messages on `medium` and `low`, the medium ones *will* be processed, but then it will reset and look once again for new `high` messages, instead of first processing any `low` messages.

Reply:
There is a flaw here (I think). 

T0: `medium` got one million messages. 
T1: We check `high`, no messages. So we moving on 
T2: We starting to consume the messages in `medium`
T5: We add one messages to `high`
T100: We still handing messages in `medium`...
T101: One million more messages are added in `medium`
T1000: We still handing messages in `medium`...
T∞: We handle the message in `high`.

Reply:
We maybe should introduce a limit, say we only handle X messages from `medium` before rechecking `high`

Reply:
Indeed. So, this "may" not be a problem ;). And it might be an issue of documentation on the `ReceiverInterface::get()` method. The idea is that you might return a "few" messages, but never all of them. This was motivated by Google PubSub, where you *can* just ask for one message, but I guess you normally "pull" more than 1: http://googleapis.github.io/google-cloud-php/#/docs/cloud-pubsub/v1.9.0/pubsub/subscription?method=pull

The intention is more to return just the one immediate message... or "few" immediate messages. Another example where this will come up is when we support multiple queues for one AMQP transport. In `ReceiverInterface::send()`, the cleanest implementation is to loop over  ALL the queues each time and return the "one or zero" messages from each. But that means `get()` would still max out at returning a number of messages equal to your number of queues, which will not be a HUGE number. 

I'm going to attempt to solve this with some PHPdoc.

New addition to the phpdoc above `ReceiverInterface::get()`

```php
/**
 * ... 
 * While this method could return an unlimited number of messages,
 * the intention is that it returns only one, or a "small number"
 * of messages each time. This gives the user more flexibility:
 * they can finish processing the one (or "small number") of messages
 * from this receiver and move on to check other receivers for messages.
 * If a this method returns too many messages, it could cause a
 * blocking effect where handling the messages received from one
 * call to get() takes a long time, blocking other receivers from
 * being called.
 * ...
```

Reply:
The alternative, as you suggested, is to (in `Worker`) only process some "limit" of messages from each Receiver before stopping. That should work fine in a system that properly supports `ack()`... which is definitely the normal use-case. I'm just not sure if there will be some case where some weird transport considers the messages we received as, sort of, auto-acked, and so if we just "drop" them in loop above this code, they will never be redelivered.